만국박람회 [STEP 1] Brad, Groot

bradheo65 opened the pull request:
안녕하세요! @ryan-son 

첫번째 Step01 PR 보내드립니다. 잘 부탁 드립니다 !!

# Step1 PR 

## 📑 배경
Json 파일에 있는 데이터를 앱에서 사용하기 위한 준비작업
## 🧑🏻‍💻 작업 내용
### 1. `ExpoInformation`
exposition_universelle_1900.json 파일 `JsonDecode`를 위한 데이터 포멧 타입 구현. 
### 2. `KoreaItems`
items.json 파일 `JsonDecode`를 위한 데이터 포멧 타입 구현. 
## 💡 테스트 방법
Unit Test에 작성된 각 Test 함수를 실행 후 결과 확인
## ✏️ 리뷰 노트
### 1. JsonDecode 방식
현재 저희가 작업한 `Decoding` 방식이 현업에서 사용하는 방식과 유사한지에 대해 의문이 있습니다.

---
### 2. `codingKey` 왜 써야 하는지??
"변수나 상수명을 스네이크 케이스로 선언해주면 동일하지 않을까"라고 생각했지만, 그렇게 되면 코드 컨벤션이 안맞기 때문에 `codingKey`를 사용해서 json `key`값에 대해 대응해주는 것 같습니다. 라이언의 생각은 어떤지 궁금합니다!!

---
### 3. Json 파일을 어디에 저장해야 할지 
- 프로젝트 파일에 저장??
- Assets에 import??
#### 결론: Assets에 images, data 등 프로젝트에서 사용하는 파일을 넣는다는 [Asset Management](https://developer.apple.com/documentation/xcode/asset-management) 문서의 내용을 보고 json 파일도 data 라고 생각해서 Assets에 import하는 방법을 채택했습니다. 라이언은 프로젝트에서 사용하는 파일을 어떤 방식으로 관리하시는지 궁금합니다.

---
### 4. Json decode 테스트 nil 반환
ExpoInformation 구조체를 구현하고 `JsonDecode`를 하는 과정에서 `nil`이 반환되는 문제가 있었습니다. 그래서 아래의 순서대로 디버깅을 해보았습니다.
1. Assets에서 Json 파일을 가져오는게 문제인지 LLDB를 활용해 확인했지만, 파일을 가져오고 있었습니다.
2. 직접 Json 데이터를 작성해서 테스트 하는 방법을 사용해서 확인했으나, 동일하게 nil을 반환했습니다.
3. `ExpoInformation` 구조체의 프로퍼티를 확인하는 과정에서 특정 Key에 오타가 있었음을 알 수 있었습니다.
#### 결론: 하나의 Key 값에 문제가 있어도 Decode 전체에 문제가 생긴다는 걸 알 수 있었습니다.

bradheo65 committed:
docs: Add expo_asests

bradheo65 committed:
feat: Create ExpoInformation Type

- Set Property
- Implement Method
- Add Unit Test

Groot-94 committed:
feat: Create KoreaItems Type

- Set Property
- Implement Method
- Add Unit Test

bradheo65 committed:
chore: Modify Creater name

Groot-94 committed:
docs: Update REAMDE.md

ryan-son commented on Expo1900/Expo1900/Model/ExpoInformation.swift:
```diff
@@ -0,0 +1,35 @@
+    
+    func getDescription() -> String {
+        return description
+    }
```
프로퍼티를 읽어오기 위해 별도의 메서드를 작성할 수 있지만, 프로퍼티에 읽기 전용의 접근 수준을 지정하면 동일한 효과를 낼 수 있을 것 같아요. 여러분은 어떻게 생각하시나요?

- [Access Control - Swift Programming Language](https://docs.swift.org/swift-book/LanguageGuide/AccessControl.html#ID18)

Groot-94 replied:
private(set)으로 수정했습니다. 
[f3efa97](https://github.com/yagom-academy/ios-exposition-universelle/pull/173/commits/f3efa97864b72e9cc3facf0f70aca85e3996bc46)

func을 선택한 이유: private(set)과 func 중 어떤 방식을 사용할 지 논의를 했었습니다. 기능의 차이가 없기 때문에 func이 좀 더 익숙하고 표현방식이 더 명확하다고 생각해서 func으로 만들었습니다.

`궁금증: 라이언의 의견을 듣고 생각해보니 func으로 만드는 경우에 메모리가 더 커질수도 있겠다고 생각이 들었습니다. 실제로 이렇게 동일한 기능을 func으로 구현하면 메모리가 더 커지는게 맞는지 궁금합니다.`

ryan-son replied:
어떤 방식으로도 접근할 수 있을 것 같네요. 하지만 이 경우는 문법적으로 기능을 지원하고, 지원하는 방법을 사용하면 값을 읽어오기 위해 별도의 메서드를 만드는 보일러 플레이트 코드를 줄일 수 있다는 이점이 있다고 생각합니다. 언어의 특성에 맞게 내용을 작성한다고도 생각해볼 수 있겠네요. 더 살펴봐야겠지만 `private(set)` 과 메서드 구현이 메모리 사용에 유의미한 차이가 있지는 않을거라고 생각해요.

ryan-son commented on Expo1900/Expo1900/Model/KoreaItems.swift:
```diff
@@ -0,0 +1,41 @@
+
+import UIKit
+
+struct KoreaItems: Decodable {
```
프로퍼티 이름을 토대로 추정해보면 이 모델은 하나의 출품작을 의미하는 것으로 추측되는데, 그럼 단수형으로 표현하는 것이 좋을 것 같네요.

bradheo65 replied:
`KoreaItems` -> `KoreaItem` 단수형으로 변경했습니다
[8fc1f6e](https://github.com/yagom-academy/ios-exposition-universelle/pull/173/commits/8fc1f6e6ec9e029302c3b8a7cce5dabea499bcb3)

처음엔 각 아이템(직지심체요절, 불경 등) 여러 아이템을 가져오는 기능과 json파일명이 `items.json`이라고 생각해서 복수형으로 작성했습니다. 다시 생각해보니 `KoreaItem`은 각 아이템들을 decoding하는 타입이기 때문에 복수형으로 쓸 이유는 없다고 생각합니다.

ryan-son replied:
네, 복수형 이름은 `KoreaItem`의 `Collection` 타입에 적용해볼 수 있을 것 같습니다.

ryan-son commented on Expo1900/Expo1900/Model/KoreaItems.swift:
```diff
@@ -0,0 +1,41 @@
+    private var name: String
+    private var imageName: String
+    private var shortDesc: String
+    private var desc: String
```
저는 팀 내부적으로 약속하지 않았다면 당연한 것이라도 이름을 줄여서 표현하지는 않아요. 누군가에겐 당연한 것이 다른 사람에게는 아닐 수 있으니까요. 이번 프로젝트에서는 최대한 줄임 표현을 사용하지 않을 수 있다면 좋겠어요. 프로퍼티의 이름을 변경하면 `CodingKeys`에서 따로 해주어야 할 일이 있을까요?

bradheo65 replied:
`shortDesc' -> shortDescription`, `desc` -> `description` 으로 변경 했습니다.
[8fc1f6e](https://github.com/yagom-academy/ios-exposition-universelle/pull/173/commits/8fc1f6e6ec9e029302c3b8a7cce5dabea499bcb3)

네이밍 작업 시 json파일에 있는 이름과 동일하게 작성해주는 것이 "json파일과 비교할 때 이해가 더 잘 될 것 같다" 라는 생각을 했었는데, 다시 보니 "그럼 스네이크케이스도 동일하게 사용해주지 않나"라고 생각도 들었습니다. 
`결론적으로 deconding은 json형식을 swift언어로 변경을 해주는 기능이기 때문에 각 json에 대한 이름들도 swift언어 형식에 맞게 작성하는게 좋을 것 같습니다.`
swift는 줄임 표현의 사용을 자제하는 것이 좋다고 되어 있어 최대한 줄임 표현을 사용하지 않도록 해보겠습니다.

ryan-son replied:
Swift 사용자들은 팀에서 합의된 내용이 아니라면 줄임 표현을 쓰는 것을 선호하지 않더라구요. 줄임 표현을 선호하지 않는다는 것에 대한 근거는 [여기](https://www.swift.org/documentation/api-design-guidelines/#use-terminology-well)에서 살펴보실 수 있습니다.

ryan-son commented on Expo1900/Expo1900/Model/KoreaItems.swift:
```diff
@@ -0,0 +1,41 @@
+    
+    func getImage() -> UIImage? {
+        return UIImage(named: imageName)
+    }
```
팀원과 협의하기에 따라 달라질 수 있지만,  Swift에서는 일반적으로 get을 생략하여 표현하는 경우가 많았던 것 같아요. `Int.random(in:)`이 `Int.getRandomNumber(in:)`이 아닌 것처럼요. 혹시 애플이 제공하는 프레임워크에서 `get~` 형식의 이름을 가진 메서드를 보신 적이 있으셨나요? 저도 궁금하네요!

Groot-94 replied:
함수를 사용하지 않는 방향으로 수정했습니다.
[8fc1f6e](https://github.com/yagom-academy/ios-exposition-universelle/pull/173/commits/8fc1f6e6ec9e029302c3b8a7cce5dabea499bcb3)

생각해보니 `get` 형식의 함수를 본 기억이 없네요..

특정 값을 가져오는 함수를 만든다고 할 때 라이언은 `get`이 아닌 `gain, acquire, obtain` 과 같은 네임을 사용하시는 지, 그게 아니라면 어떤 방식의 네이밍을 하시는 지 궁금합니다!

ryan-son replied:
값을 가져오는 경우 저는 명사형의 이름을 사용하는 것을 선호하고, 네트워크를 통해 데이터를 가져올 때는 `request`, 기기 디스크로부터 가져올 때는 `fetch`와 같은 이름을 사용하고 있어요. 물론 취향이 반영된겁니다!

ryan-son commented on Expo1900/ExpoInformationTest/ExpoInformationTest.swift:
```diff
@@ -0,0 +1,81 @@
+    
+    override func setUpWithError() throws {
+        try super.setUpWithError()
+        expo1900 = NSDataAsset.init(name: "exposition_universelle_1900")
```
이 구문에서 명시적으로 `.init`을 작성하신 이유가 궁금하네요. `UIImage(name:)`과 같은 요소들은 `UIImage.init(name:)`라고 작성하지 않았으니 생성자 호출 방식을 하나의 방식으로 통일하는 것이 더 좋을 것 같네요. 어떤 쪽이 더 일반적인 표현 방식일까요?

Groot-94 replied:
수정했습니다.
[91a44f5](https://github.com/yagom-academy/ios-exposition-universelle/pull/173/commits/91a44f505c8e816178afa1ea8ec5e65baeeddce3)

NSDataAsset에 대해 찾아보면서 외부에서 사용하는 방식 그대로 사용했습니다.
`UIImage(name:)`으로 사용하는 방법이 더 일반적이라고 생각이 됩니다.

ryan-son commented on Expo1900/ExpoInformationTest/ExpoInformationTest.swift:
```diff
@@ -0,0 +1,81 @@
+        
+        // then
+        XCTAssertEqual(result, expoInformation?.getDescription().prefix(10).description)
+    }
```
이 테스트들은 모두 디코딩 성공 여부에 관심이 있으니 하나의 테스트로 합칠 수 있을 것 같아요. 지금과 같이 Assert 구문에서 title, visitor 등을 개별적으로 확인한다면 만약 실패하더라도 어떤 부분에서 실패했는지 바로 확인할 수 있으니 효과는 다르지 않을 것 같네요.

Groot-94 replied:
title만 테스트 하는 방법으로 수정했습니다.
[91a44f5](https://github.com/yagom-academy/ios-exposition-universelle/pull/173/commits/91a44f505c8e816178afa1ea8ec5e65baeeddce3)

다시 생각해보니 `test_exposition_universelle_1900파일의_title값이_디코딩되는지()` 만으로도 충분히 전체적인 decoding의 문제를 파악 할 수 있다고 생각이 들었습니다.

ryan-son replied:
동의합니다. 추가로 JSON이 Asset에 없거나, JSON의 이름이 맞지 않아 가져오지 못하는 경우를 명확하게 표현하기 위해 setUp 단계에서 `NSAsset`을 가져올 때 옵셔널 바인딩과 [`preconditionFailure(_:)`](https://developer.apple.com/documentation/swift/preconditionfailure(_:file:line:)) 구문을 추가해보는 것도 좋을 것 같네요.

ryan-son commented on Expo1900/KoreaItemsTest/KoreaItemsTest.swift:
```diff
@@ -0,0 +1,82 @@
+        let result = "직지심체요절"
+        
+        //then
+        XCTAssertEqual(result, itemInformation![0].getName())
```
아래와 같이 변경하면 테스트가 동작할까요? 왜 그럴까요?
```suggestion
        XCTAssertEqual(result, itemInformation?[0].getName())
```

Groot-94 replied:
동작을 확인했습니다.

[91a44f5](https://github.com/yagom-academy/ios-exposition-universelle/pull/173/commits/91a44f505c8e816178afa1ea8ec5e65baeeddce3)

XCTAssertEqual 함수도 `==` 연산자를 사용해서 비교해주기 때문에 ! 를 사용해서 옵셔널을 제거해주는 작업이 불필요하다고 생각이 듭니다.
`아래와 같이 변경하면 테스트가 동작할까요? 왜 그럴까요?`라고 물어보시는 의도를 저희가 파악을 제대로 했나요??

ryan-son replied:
`==` 연산자를 통한 비교 과정에서 `result`는 옵셔널이 아니고, `itemInformation?[0].getName()`은 옵셔널 값인데 어떻게 비교가 가능했을까에 대해 여쭈어보려고 했었습니다~! 자세한 내용은 이 문서를 참고해보세요~ [==(_:_:)](https://developer.apple.com/documentation/swift/optional/==(_:_:)-m6x)

ryan-son commented on Expo1900/KoreaItemsTest/KoreaItemsTest.swift:
```diff
@@ -0,0 +1,82 @@
+        
+        //then
+        XCTAssertEqual(result, itemInformation![0].getDesc().prefix(10).description)
+    }
```
이 테스트들도 하나로 합칠 수 있을 것 같네요. 여러분 생각은 어떠신가요?

Groot-94 replied:
name만 테스트 하는 방법으로 수정했습니다.
[91a44f5](https://github.com/yagom-academy/ios-exposition-universelle/pull/173/commits/91a44f505c8e816178afa1ea8ec5e65baeeddce3)

ryan-son commented on README.md:
```diff
@@ -1,6 +1,143 @@
+    - [고민한점](#STEP-1-고민한점)
+    - [배운 개념](#STEP-1-배운개념)
+    - [Review](#STEP-1-Review)
+    - [Update](#STEP-1-Update)
```
Readme를 바로 작성하시다니 부지런하시네요. 멋집니다! 👏

ryan-son commented on README.md:
```diff
@@ -1,6 +1,143 @@
+---
+## 1️⃣ STEP 1
+### STEP 1 고민한점
+- json Decoding 방식
+- coding key 왜 써야 하는지??
+- Json decode 테스트 nil 반환
```
앞으로 고민한 내용에 대한 결과도 작성되면 내용이 더 알찰 것 같네요~! snake case와 camel case 간 간단한 변환 같은 경우 [keyDecodingStrategy](https://developer.apple.com/documentation/foundation/jsondecoder/2949119-keydecodingstrategy)를 통해 수행할 수 있으니 참고하세요~

Groot-94 replied:
[keyDecodingStrategy](https://developer.apple.com/documentation/foundation/jsondecoder/2949119-keydecodingstrategy)를 통해서 Json 형식의 많은 Key 값을 한 번에 원하는 방향으로 수정할 수 있다고 생각이 들었습니다. 
좋은 내용 감사합니다!

ryan-son commented on Expo1900/Expo1900.xcodeproj/project.pbxproj:
```diff
@@ -70,6 +141,42 @@
+			productName = ExpoInformationTest;
+			productReference = A3DF407728575B2F005B1FC9 /* ExpoInformationTest.xctest */;
+			productType = "com.apple.product-type.bundle.unit-test";
+		};
```
다른 모델이지만 앱 안에서 동일한 시점에 이용할 모델이라 테스트 타겟을 별도로 나누지는 않아도 될 것 같아요. 테스트 파일을 하나의 타겟 안에 구성하는건 어떨까요?

bradheo65 replied:
테스트 타겟을 통합해서 사용해 보았습니다.
[51c6e82](https://github.com/yagom-academy/ios-exposition-universelle/pull/173/commits/51c6e82b197c2fff8e9467623ce089ca4dd299a8)

ryan-son reviewed (changes requested):
브래드, 그루트, 반갑습니다! Step 1 수행하시느라 고생 많으셨어요. 먼저 PR 본문을 통해 질문 주신 내용부터 답변드릴게요.

> ### 1. JsonDecode 방식
> 현재 저희가 작업한 `Decoding` 방식이 현업에서 사용하는 방식과 유사한지에 대해 의문이 있습니다.

JSON 형식의 데이터를 받는다면 여러분이 정의하신 것과 같은 방식으로 `Decodable`을 준수하는 모델을 정의하고, `JSONDecoder`를 이용해 디코딩하는 방식을 사용하고 있어요.

---
>### 2. `codingKey` 왜 써야 하는지??
>"변수나 상수명을 스네이크 케이스로 선언해주면 동일하지 않을까"라고 생각했지만, 그렇게 되면 코드 컨벤션이 안맞기 때문에 `codingKey`를 사용해서 json `key`값에 대해 대응해주는 것 같습니다. 라이언의 생각은 어떤지 궁금합니다!!

전달받은 JSON의 키값과 정의한 모델의 프로퍼티의 이름을 다르게 만들고 싶은 경우가 있을 수 있어요. 그럴 때 `CodingKey` 프로토콜을 준수하는 `CodingKeys` 열거형을 정의합니다. 단순히 Snake case 형식의 키값을 Camel case로 변경하는 경우에는 `JSONDecoder`의 [`keyDecodingStrategy`](https://developer.apple.com/documentation/foundation/jsondecoder/2949119-keydecodingstrategy)를 변경하여 대응할 수도 있으니 참고하세요.

---
>### 3. Json 파일을 어디에 저장해야 할지 
> - 프로젝트 파일에 저장??
> - Assets에 import??
> #### 결론: Assets에 images, data 등 프로젝트에서 사용하는 파일을 넣는다는 [Asset Management](https://developer.apple.com/documentation/xcode/asset-management) 문서의 내용을 보고 json 파일도 data 라고 생각해서 Assets에 import하는 방법을 채택했습니다. 라이언은 프로젝트에서 사용하는 파일을 어떤 방식으로 관리하시는지 궁금합니다.

일반적으로 JSON 형식의 파일을 저장하는 경우는 드물긴 하지만 앱에서 활용한다면 Asset에 위치시킬 것 같아요. 하지만 단순히 테스트를 위해 활용하는 JSON 형식의 데이터가 필요하다면 별도로 마련한 테스트 타겟에서 상수로 정의하여 사용할거에요. 현재 Asset 위치에 넣는다면 불필요하게 앱 용량이 커질테니까요.

고생하셨습니다! 이해되지 않는 코멘트가 있으시거나 궁금한 내용이 있으시면 바로 DM 주세요~

Groot-94 committed:
refactor: Modify ExpoInformation

- Remove return functions of property
- Modify access control of property

bradheo65 committed:
refactor: Modify KoreaItem

- Modify KoreaItem name (KoreaItems -> KoreaItem)
- Modify name of property
- Remove return functions of property
- Modify access control of property

Groot-94 committed:
refactor: Modify Unit Test

- Merge Decoding Test (ExpoInformationTest + KoreaItemTest)
- Modify initialization of NSDataAsset
- Modify Optional set (Forced Unwrapping -> Optional Chaining)

bradheo65 committed:
refactor: Merge Target

- Merge Unit Test Target

Groot-94 committed:
docs: Update REAMDE.md

Groot-94 committed:
docs: Update REAMDE.md

- Update UML

ryan-son commented on Expo1900/Expo1900/Model/KoreaItem.swift:
```diff
@@ -0,0 +1,26 @@
+//  Created by Brad, Groot on 2022/06/14.
+//
+
+import UIKit
```
`import UIKit.UIImage`와 같이 특정 submodule을 import 할 수 있는 방법도 있어요. 이처럼 해당 모듈의 특정 class, struct, typealias 등만 import하여 이미 소스 내에서 정의된 내용과 충돌을 피하는 방법이 있으니 참고하세요~ [Import Declaration - Swift Programming Language](https://docs.swift.org/swift-book/ReferenceManual/Declarations.html#ID354)

ryan-son reviewed (approved):
고생 많으셨습니다!

커밋 메세지를 남기실 때 제목과 같은 역할을 하는 첫 라인에 어떤 변경사항이 있었는지 요약해서 작성해주시면 더 좋을 것 같네요. 그럼 다음 스텝에서 뵙겠습니다!
![image](https://user-images.githubusercontent.com/69730931/173983315-90a6b0b8-e412-40f2-9b2e-d1e2ee45bc38.png)
